From a49fcfbff25dc86d91bbd45a2b2d5d5e4d35cf58 Mon Sep 17 00:00:00 2001
From: Aleksandr Mezin <mezin.alexander@gmail.com>
Date: Fri, 9 Oct 2026 06:51:47 +0300
Subject: [PATCH] coverage: Fix source file filtering

`SF:` lines generated by `GetCodeCoverageSummary()` usually contain
URLs (`resource://` and `file://`), but they were passed to
`g_canonicalize_filename()`, as if they were file names,
and then compared to similarly canonicalized prefixes
(potentially also URLs), which produced unpredictable results:

- Internal GJS sources included in `coverage.lcov` and written
to the working directory - even though there was no matching
`--coverage-prefix`.

- Modules imported from the file system were incorrectly filtered
out, even though they were imported from `--coverage-prefix`.

- Also, `g_canonicalize_filename()` discards the trailing slash.
So `/prefix/` surprisingly matched `/prefixabc`.

Instead, convert everything to GFiles, and compare them using
`g_file_has_prefix()` and `g_file_equal()`.

Backport of https://gitlab.gnome.org/GNOME/gjs/-/merge_requests/1127
to v51 branch.
---
 gjs/coverage.cpp           | 44 +++++++++++++++++-----------------
 test/gjs-test-coverage.cpp | 48 ++++++++++++++++++++++++++++++++++++++
 2 files changed, 69 insertions(+), 23 deletions(-)

diff --git a/gjs/coverage.cpp b/gjs/coverage.cpp
index c43acc734..f78809880 100644
--- a/gjs/coverage.cpp
+++ b/gjs/coverage.cpp
@@ -8,8 +8,10 @@
 #include <stdint.h>
 #include <string.h>  // for strcmp, strlen
 
+#include <algorithm>
 #include <iterator>  // for size
 #include <new>
+#include <vector>
 
 #include <gio/gio.h>
 #include <glib-object.h>
@@ -160,22 +162,6 @@ static Gjs::AutoChar find_diverging_child_components(GFile* child,
     return strip_uri_scheme(child_uri);
 }
 
-[[nodiscard]]
-static bool filename_has_coverage_prefixes(GjsCoverage* self,
-                                           const char* filename) {
-    auto* priv = static_cast<GjsCoveragePrivate*>(
-        gjs_coverage_get_instance_private(self));
-    Gjs::AutoChar workdir{g_get_current_dir()};
-    Gjs::AutoChar abs_filename{g_canonicalize_filename(filename, workdir)};
-
-    for (const char* const* prefix = priv->prefixes; *prefix; prefix++) {
-        Gjs::AutoChar abs_prefix{g_canonicalize_filename(*prefix, workdir)};
-        if (g_str_has_prefix(abs_filename, abs_prefix))
-            return true;
-    }
-    return false;
-}
-
 [[nodiscard]]
 static inline GErrorResult<> write_line(GOutputStream* out, const char* line) {
     Gjs::AutoError error;
@@ -221,6 +207,11 @@ GErrorResult<Gjs::AutoUnref<GFile>> write_statistics_internal(GjsCoverage* self,
     if (!ostream)
         return Err(error.release());
 
+    std::vector<Gjs::AutoUnref<GFile>> prefixes;
+
+    for (const char* const* prefix = priv->prefixes; *prefix; prefix++)
+        prefixes.emplace_back(g_file_new_for_commandline_arg(*prefix));
+
     Gjs::AutoStrv lcov_lines{g_strsplit(lcov.get(), "\n", -1)};
     const char* test_name = nullptr;
     bool ignoring_file = false;
@@ -241,19 +232,26 @@ GErrorResult<Gjs::AutoUnref<GFile>> write_statistics_internal(GjsCoverage* self,
 
         if (g_str_has_prefix(*iter, "SF:")) {
             const char* filename = *iter + 3;
-            if (!filename_has_coverage_prefixes(self, filename)) {
-                ignoring_file = true;
-                continue;
-            }
-
-            // Now we can write the test name before writing the source file
-            MOZ_TRY(write_line(ostream, test_name));
 
             /* The source file could be a resource, so we must use
              * g_file_new_for_commandline_arg() to disambiguate between URIs and
              * filesystem paths. */
             Gjs::AutoUnref<GFile> source_file{
                 g_file_new_for_commandline_arg(filename)};
+
+            ignoring_file = std::none_of(
+                prefixes.begin(), prefixes.end(),
+                [&source_file](GFile* prefix) {
+                    return g_file_has_prefix(source_file, prefix) ||
+                           g_file_equal(source_file, prefix);
+                });
+
+            if (ignoring_file)
+                continue;
+
+            // Now we can write the test name before writing the source file
+            MOZ_TRY(write_line(ostream, test_name));
+
             Gjs::AutoChar diverged_paths{
                 find_diverging_child_components(source_file, priv->output_dir)};
             Gjs::AutoUnref<GFile> destination_file{
diff --git a/test/gjs-test-coverage.cpp b/test/gjs-test-coverage.cpp
index 2a97b333b..deaa8e2bc 100644
--- a/test/gjs-test-coverage.cpp
+++ b/test/gjs-test-coverage.cpp
@@ -370,6 +370,47 @@ static void test_expected_entry_not_written_for_nonexistent_file(
     g_assert_null(sf_line);
 }
 
+static void test_expected_entry_written_for_module_evaluated_from_file(
+    void* fixture_data, const void*) {
+    auto* fixture = static_cast<GjsCoverageFixture*>(fixture_data);
+
+    Gjs::AutoChar filename{get_script_identifier(fixture->tmp_js_script)};
+    bool ok = gjs_context_eval_module_file(fixture->gjs_context, filename,
+                                           nullptr, nullptr);
+    g_assert_true(ok);
+
+    Gjs::AutoChar coverage_data_contents =
+        write_statistics_and_get_coverage_data(fixture->coverage,
+                                               fixture->lcov_output);
+
+    Gjs::AutoChar expected_source_filename = get_output_path_for_script_on_disk(
+        fixture->tmp_js_script, fixture->lcov_output_dir);
+
+    assert_coverage_data_contains_value_for_key(
+        coverage_data_contents, "SF:", expected_source_filename);
+}
+
+static void test_entry_not_written_for_internal_sources(void* fixture_data,
+                                                        const void*) {
+    auto* fixture = static_cast<GjsCoverageFixture*>(fixture_data);
+
+    Gjs::AutoChar workdir{g_get_current_dir()};
+    char* coverage_paths[] = {workdir, nullptr};
+
+    g_object_unref(fixture->coverage);
+    fixture->coverage = gjs_coverage_new(coverage_paths, fixture->gjs_context,
+                                         fixture->lcov_output_dir);
+
+    /* Creating the context already executed internal bootstrap scripts, no
+     * need to evaluate anything else */
+    Gjs::AutoChar coverage_data_contents =
+        write_statistics_and_get_coverage_data(fixture->coverage,
+                                               fixture->lcov_output);
+
+    const char* sf_line = line_starting_with(coverage_data_contents, "SF:");
+    g_assert_null(sf_line);
+}
+
 enum BranchTaken : uint8_t { NOT_EXECUTED, NOT_TAKEN, TAKEN };
 
 struct BranchLineData {
@@ -1058,6 +1099,13 @@ void gjs_test_add_tests_for_coverage() {
     add_test_for_fixture("/gjs/coverage/entry_not_written_for_nonexistent_file",
                          &coverage_fixture,
                          test_expected_entry_not_written_for_nonexistent_file);
+    add_test_for_fixture(
+        "/gjs/coverage/entry_written_for_module_evaluated_from_file",
+        &coverage_fixture,
+        test_expected_entry_written_for_module_evaluated_from_file);
+    add_test_for_fixture("/gjs/coverage/entry_not_written_for_internal_sources",
+                         &coverage_fixture,
+                         test_entry_not_written_for_internal_sources);
     add_test_for_fixture(
         "/gjs/coverage/single_branch_coverage_written_to_coverage_data",
         &coverage_fixture,
-- 
GitLab

